File: 052_BinaryTreeLevelOrderTraversal.cpp
// Topic: Binary Tree Level Order Traversal
// Description: Traverse tree by level using BFS and return nodes grouped by level.
// Approach: Standard BFS using queue.

#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* buildLevel(const vector<string>& vals)
{
    if (vals.empty() || vals[0] == "null")
        return nullptr;
    TreeNode* root = new TreeNode(stoi(vals[0]));
    queue<TreeNode*> q;
    q.push(root);
    int i = 1;
    while (!q.empty() && i < (int)vals.size())
    {
        TreeNode* node = q.front();
        q.pop();
        if (i < (int)vals.size() && vals[i] != "null")
        {
            node->left = new TreeNode(stoi(vals[i]));
            q.push(node->left);
        }
        ++i;
        if (i < (int)vals.size() && vals[i] != "null")
        {
            node->right = new TreeNode(stoi(vals[i]));
            q.push(node->right);
        }
        ++i;
    }
    return root;
}

vector<vector<int>> levelOrder(TreeNode* root)
{
    vector<vector<int>> res;
    if (!root)
        return res;
    queue<TreeNode*> q;
    q.push(root);
    while (!q.empty())
    {
        int sz = q.size();
        vector<int> level;
        level.reserve(sz);
        while (sz--)
        {
            TreeNode* node = q.front();
            q.pop();
            level.push_back(node->val);
            if (node->left)
                q.push(node->left);
            if (node->right)
                q.push(node->right);
        }
        res.push_back(move(level));
    }
    return res;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    vector<string> vals = {"3", "9", "20", "null", "null", "15", "7"};
    TreeNode* root = buildLevel(vals);
    auto out = levelOrder(root);
    for (auto& lvl : out)
    {
        for (size_t i = 0; i < lvl.size(); ++i)
        {
            if (i)
                cout << " ";
            cout << lvl[i];
        }
        cout << "\n";
    }
}

/*
Time Complexity: O(n)
Space Complexity: O(n)
*/

File: 052_BinaryTreeLevelOrderTraversal.py
# Topic: Binary Tree Level Order Traversal
# Description: Traverse tree by level using BFS and return nodes grouped by level.
# Approach: Standard BFS using queue.

from typing import Optional, List
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        result = []
        q = deque([root])
        while q:
            level = []
            for _ in range(len(q)):
                node = q.popleft()
                level.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            result.append(level)
        return result


if __name__ == "__main__":
    # Test 1
    root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    print(f"Test 1 -> {Solution().levelOrder(root)}")

    # Test 2
    root = None
    print(f"Test 2 -> {Solution().levelOrder(root)}")

"""
Time Complexity: O(n)
Space Complexity: O(n)
"""
